iceoryx_utils/include/iceoryx_utils/internal/units/duration.inl

```diff
@@ -20,262 +20,511 @@ namespace iox
+
+    // the m_nanoseconds multiplication cannot exceed the limits of a Duration, since m_nanoseconds is always less than
+    // a second and m_seconds can hold 64 bits and the multiplicator is at max 64 bits
+
+    // check if the result of the m_nanoseconds multiplication can easily be converted into a Duration
+    if (m_nanoseconds <= maxBeforeOverflow)
+    {
+        return durationFromSeconds + Duration::nanoseconds(m_nanoseconds * multiplicator);
+    }
+
+    // when we reach this, the multiplicator must be larger than 2^32, since smaller values multiplied with the
+    // m_nanoseconds(uint32_t) would fit into 64 bits;
+    // to accurately determine the result, the calculation is split into a multiplication with the lower 32 bits of the
+    // multiplicator and another one with the upper 32 bits;
+
+    // this is the easy part with the lower 32 bits
+    uint64_t multiplicatorLow = static_cast<uint32_t>(multiplicator);
+    auto durationFromNanosecondsLow = Duration::nanoseconds(m_nanoseconds * multiplicatorLow);
+
+    // this is the complicated part with the upper 32 bits;
+    // the m_nanoseconds are multiplied with the upper 32 bits of the multiplicator shifted by 32 bit to the left, thus
+    // having again a multiplication of two 32 bit values whose result fits into a 64 bit variable;
+    // one bit of the result represents 2^32 nanoseconds;
+    // just shifting left by 32 bits would result in an overflow, therefore blocks of full seconds must be extracted of
+    // the result;
+    // this cannot be done by dividing through NANOSECS_PER_SEC, since that one is base 1_000_000_000 and the result is
+    // base 2^32, therefore the least common multiple must be used to get blocks of full seconds represented as 2^32
+    // nanoseconds per bit;
+    // this can then safely be converted to seconds as well as nanoseconds without loosing precision
+
+    // least common multiple of 2^32 and NANOSECONDS_PER_SECOND
+    constexpr uint64_t LEAST_COMMON_MULTIPLE{8388608000000000};
+    static_assert(LEAST_COMMON_MULTIPLE % (1ULL << 32) == 0, "invalid multiple");
+    static_assert(LEAST_COMMON_MULTIPLE % NANOSECS_PER_SEC == 0, "invalid multiple");
+
+    constexpr uint64_t ONE_FULL_BLOCK_OF_SECONDS_ONLY{LEAST_COMMON_MULTIPLE >> 32};
+    constexpr uint64_t SECONDS_PER_FULL_BLOCK{LEAST_COMMON_MULTIPLE / NANOSECS_PER_SEC};
+
+    uint64_t multiplicatorHigh = static_cast<uint32_t>(multiplicator >> 32U);
+    auto nanosecondsFromHigh = m_nanoseconds * multiplicatorHigh;
+    auto fullBlocksOfSecondsOnly = nanosecondsFromHigh / ONE_FULL_BLOCK_OF_SECONDS_ONLY;
+    auto remainingBlockWithFullAndFractionalSeconds = nanosecondsFromHigh % ONE_FULL_BLOCK_OF_SECONDS_ONLY;
+
+    auto durationFromNanosecondsHigh = Duration{fullBlocksOfSecondsOnly * SECONDS_PER_FULL_BLOCK, 0U}
+                                       + Duration::nanoseconds(remainingBlockWithFullAndFractionalSeconds << 32);
+
```

Comment:
OK.
1) nanosecondsFromHigh fits into 64 bits.
2) fullBlocksOfSecondsOnly fits into 64 Bits, counts only some seconds and no leftover nanoseconds in multiples of ONE_FULL_BLOCK_OF_SECONDS_ONLY
3) remainingBlockWithFullAndFractionalSeconds fits into 64 bit and contains the remaining nanoseconds, may exceed 10⁹ i.e. contain some full seconds

4) fullBlocksOfSecondsOnly * SECONDS_PER_FULL_BLOCK should still fit into 64 bit
5) Can we guarantee that remainingBlockWithFullAndFractionalSeconds << 32 always fits into 64 bits as well?
Otherwise we might have a problem in certain edge cases.

I need to rethink this ...

If 5) causes no problem, it should be correct (operator+ deals with potential further overflows)

Reply:
yes, it is the modulo of a number less than 2^32 therefore it can safely be shifted